### PR TITLE
strumpack: Changed old test method to new test method

### LINKED ISSUE #44848: Testing issue: strumpack
### Steps to reproduce the failure(s) or link(s) to test output(s)

```console
$ spack test run strumpack

```

### Error message

<details><summary>Error message</summary>

<pre>
        232        # TODO: Remove/replace once self.spec['cmake'] is available here
        233        cmake_bin = self.cmake_bin(set=False)
        234
  >>    235        opts = self.std_cmake_args
        236        opts += self.cmake_args()
        237        opts += ["."]
        238

</pre></details>

### Information on your system or the test runner

* **Spack:** 0.23.0.dev0 (27e28b33eecbf1f47bca2675340dbaeb669827e5)
* **Python:** 3.9.12
* **Platform:** linux-rhel8-broadwell
* **Concretizer:** clingo

### Additional information
```
======================= SUMMARY: strumpack-7.2.0-ck6btof =======================
Strumpack::test .. FAILED
============================= 1 failed of 1 parts ==============================
==> [2024-06-24-16:14:51.989947]

See test results at:
/.spack/test/gscrja4vk4bpfkymzhrl3rwxrctvaehu/strumpack-7.2.0-ck6btof-test-out.txt
Test failure: TestFailure: 1 test failed.


'Strumpack' object has no attribute 'std_cmake_args'

/spack/var/spack/repos/builtin/packages/strumpack/package.py:235, in _test_example:
        232        # TODO: Remove/replace once self.spec['cmake'] is available here
        233        cmake_bin = self.cmake_bin(set=False)
        234
^[[0;91m  >>    235        opts = self.std_cmake_args^[[0m
        236        opts += self.cmake_args()
        237        opts += ["."]
        238
```

### General information

- [X] I have reported the version of Spack/Python/Platform/Runner
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded any available logs
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by AcriusWinter**
Update standalone test API. See 2nd comment for test output.

Fixes #44848 
Supersedes #35807  (for one package)

https://spack.readthedocs.io/en/latest/packaging_guide.html#stand-alone-tests

### CODE REVIEW

**Comment by tldahlgren** [comment]
There are style checks that need to be resolved:

```
+++ var/spack/repos/builtin/packages/strumpack/package.py	2024-06-25 16:26:50.315457+00:00
@@ -247,9 +247,9 @@
                 self._test_example(test_exe_mpi, exe, test_args)
                 return
             except:
                 continue
         assert False, "No MPI executable was found"
-    
+
     def check(self):
         """Skip the builtin testsuite, use the stand-alone tests instead."""
         pass
would reformat var/spack/repos/builtin/packages/strumpack/package.py
Oh no! 💥 💔 💥
1 file would be reformatted.
  black found errors
==> Running flake8 checks
var/spack/repos/builtin/packages/strumpack/package.py:249: [E722] do not use bare 'except'
var/spack/repos/builtin/packages/strumpack/package.py:252: [W293] blank line contains whitespace
  flake8 found errors
==> Running mypy checks
==> Error: spack style found errors
```

I recommend running `black` locally since the `flake8` on line 249 will have to be fixed.

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 1]
```diff
@@ -229,51 +213,43 @@ def _test_example(self, test_prog, test_dir, test_cmd, test_args):
                 "target_link_libraries({0} ".format(test_prog) + "PRIVATE STRUMPACK::strumpack)\n"
             )
 
-        # TODO: Remove/replace once self.spec['cmake'] is available here
-        cmake_bin = self.cmake_bin(set=False)
+        with working_dir(test_dir):
+            opts = self.builder.std_cmake_args
+            opts += self.cmake_args()
+            opts += ["."]
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/strumpack/package.py:219]:
```suggestion
            opts = self.builder.std_cmake_args + self.cmake_args() + ["."]
```

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 2]
```diff
@@ -229,51 +213,43 @@ def _test_example(self, test_prog, test_dir, test_cmd, test_args):
...
+                exe = which(test_cmd)
+                exe(*test_args)
 
-    def test(self):
-        """Run the stand-alone tests for the installed software."""
-        test_dir = join_path(self.test_suite.current_test_cache_dir, self.test_src_dir)
+    def test_sparse_seq(self):
+        """Run sequential test_sparse"""
+        if "+mpi" in self.spec:
+            raise SkipTest("Package must be installed with '~mpi'")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/strumpack/package.py:233]: Turns out this should not be necessary, at least not for the latest version.

```suggestion
```

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 3]
```diff
@@ -229,51 +213,43 @@ def _test_example(self, test_prog, test_dir, test_cmd, test_args):
...
+            raise SkipTest("Package must be installed with '+mpi'")
+        test_exe_mpi = "test_sparse_mpi"
+        test_args = ["-n", "1", test_exe_mpi, join_path("..", self.test_data_dir, "pde900.mtx")]
+        mpiexe_list = ["srun", "mpirun", "mpiexec"]
+        for exe in mpiexe_list:
+            try:
+                self._test_example(test_exe_mpi, exe, test_args)
+                return
+            except:
+                continue
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/strumpack/package.py]: The exception needs to be caught to pass the checks.
```suggestion
            except Exception:
                pass
```

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 4]
```diff
@@ -229,51 +213,43 @@ def _test_example(self, test_prog, test_dir, test_cmd, test_args):
...
+        assert False, "No MPI executable was found"
+
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/strumpack/package.py]:
```suggestion

```

**Review by tldahlgren — DISMISSED** [review]
This PR makes the appropriate API conversions.

**Comment by tldahlgren** [comment]
For a package with `~mpi`:

```
$ spack -v test run /5zncxhrj5mzvbg5p24ds4iwekvrdl3v3
==> Spack test vsosjktx2atl6xavpt5fgdqz6rqd7stk
==> Testing package strumpack-7.2.0-5zncxhr
...
==> [2024-06-27-18:44:40.402074] test: test_sparse_mpi: Run parallel test_sparse
SKIPPED: Strumpack::test_sparse_mpi: Package must be installed with '+mpi'
==> [2024-06-27-18:44:40.402619] test: test_sparse_seq: Run sequential test_sparse
...
# COMPONENTWISE SCALED RESIDUAL = 3.21559e-16
# RELATIVE ERROR = 4.80895e-16
PASSED: Strumpack::test_sparse_seq
==> [2024-06-27-18:44:47.716998] Completed testing
==> [2024-06-27-18:44:47.717134] 
======================= SUMMARY: strumpack-7.2.0-5zncxhr =======================
Strumpack::test_sparse_mpi .. SKIPPED
Strumpack::test_sparse_seq .. PASSED
======================== 1 skipped, 1 passed of 2 parts ========================
============================== 1 passed of 1 spec ==============================
```

**Comment by tldahlgren** [comment]
I guess that style fix didn't work as expected.

@spackbot fix style

**Review by pghysels — APPROVED** [review]

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**